modules/Emails/EmailFromValidator.php

```diff
@@ -0,0 +1,470 @@
+    const ERR_FIELD_FROM_ADDR_NAME_IS_INVALID = 15;
+    const ERR_FIELD_FROM_ADDR_NAME_DOESNT_MATCH_REGEX = 16;
+    const ERR_FIELD_FROM_ADDR_NAME_INVALID_NAME_PART = 17;
+    const ERR_FIELD_FROM_ADDR_NAME_INVALID_EMAIL_PART = 18;
+    const ERR_FIELD_FROM_ADDR_NAME_INVALID_EMAIL_PART_TO_FIELD_FROM = 19;
+    const ERR_FIELD_FROM_ADDR_NAME_INVALID_EMAIL_PART_TO_FIELD_FROM_ADDR = 20;
+    const ERR_FIELD_FROM_ADDR_NAME_INVALID_EMAIL_PART_TO_FIELD_FROMNAME = 21;
+    const ERR_FIELD_FROM_ADDR_NAME_INVALID_EMAIL_PART_TO_FIELD_FROM_NAME = 22;
+    
+    const EX_ERROR_CODE_TYRE_IS_INCORRECT = 100;
+    const EX_ERROR_CODE_IS_NOT_IMPLEMENTED = 101;
+    
+    /**
+     *
+     * @var Email;
+     */
+    protected $email;
+    
+    /**
+     *
+     * @var array
+     */
+    protected $errors;
+    
+    
+    /**
+     * Specially use before email sending.
+     *
+     * @param Email $email
+     * @param bool $tryToFix
+     * @return array
+     */
+    public function isValid(Email $email, $tryToFix = true)
+    {
+        $this->setEmail($email);
+        $this->clearErrors();
+        $this->addErrors($this->validateFrom());
+        $this->addErrors($this->validateFromAddr());
+        $this->addErrors($this->validateFromName());
+        $this->addErrors($this->validateFrom_Name());
+        $this->addErrors($this->validateFromAddrName());
+        // Should be finished the rest fields like to, to address, cc, bcc etc.. (reply-to??)
+//        $this->addErrors($this->validateTo());
+//        $this->addErrors($this->validateCCs());
+//        $this->addErrors($this->validateBCCs());
+        $valid = !$this->hasErrors();
+        if (!$valid && $tryToFix) {
+            $valid = !$this->hasErrors();
+        }
+        return $valid;
+    }
+    
+    /**
+     *
+     * @param Email $email
+     */
+    protected function setEmail(Email $email)
+    {
+        $this->email = $email;
+    }
+    
+    /**
+     *
+     * @return Email
+     */
+    protected function getEmail()
+    {
+        if (!$this->email) {
+            throw new EmailValidatorException('Trying to get Email but previously is not set.',
+                EmailValidatorException::EMAIL_IS_NOT_SET);
+        }
+        if (!($this->email instanceof Email)) {
+            throw new EmailValidatorException('Trying to get Email but object type is incorrect:' . gettype($this->email),
+                EmailValidatorException::EMAIL_ISNT_EMAILOBJ);
+        }
+        return $this->email;
+    }
+    
+    /**
+     *
+     */
+    protected function clearErrors()
+    {
+        $this->errors = [];
+    }
+    
+    /**
+     *
+     * @param int $error
+     */
+    protected function addError($error)
+    {
+        if ($error !== (int)$error) {
+            throw new InvalidArgumentException('Error code should be an integer, ' . gettype($error) . ' given', self::EX_ERROR_CODE_TYRE_IS_INCORRECT);
+        }
+        if (!in_array($error, $this->errors)) {
```

Comment:
Missing type safety param

Reply:
I don't understand this request. `$this->errors` is set in constructor. Can not be anything else, only an array.